File: src/com/charter/enterprise/collatz/util/Collatz.java
package com.charter.enterprise.collatz.util;

import com.charter.enterprise.collatz.model.CollatzDataModel;

import java.util.concurrent.Callable;

public class Collatz implements Callable<CollatzDataModel> {

    CollatzDataModel collatzDataModel;

    OddEvenDeterminer oddEvenDeterminer = new OddEvenDeterminer();

    public Collatz(long startingNumber) {
        this.collatzDataModel = new CollatzDataModel(startingNumber, 0);
    }

    public CollatzDataModel getCollatzInformation(long startingNumber) {
        collatzDataModel.setCollatzNumber(startingNumber);
        collatzDataModel.setCollatzIterations(getCollatzIterations(startingNumber));

        return new CollatzDataModel(collatzDataModel);
    }

    public CollatzDataModel call() {
        return getCollatzInformation(collatzDataModel.getCollatzNumber());
    }

    public long getCollatzIterations(long startingNumber){
        long collatzIterations = 0;
        long collatzNumber = startingNumber;

        while (collatzNumber > 1) {
            if (oddEvenDeterminer.isOdd(collatzNumber)) {
                collatzIterations++;
                collatzNumber = collatzOddIteration(collatzNumber);
            } else {
                collatzIterations++;
                collatzNumber = collatzEvenIteration(collatzNumber);
            }
        }

        return collatzIterations;
    }

    public long collatzOddIteration(long collatzNumber) {
        return ((collatzNumber * 3) + 1);
    }

    public long collatzEvenIteration(long collatzNumber) {
        return (collatzNumber / 2);
    }
}

File: src/util/collatz.ts
import { CollatzDataModel } from '../model/collatz-data-model';
import { OddEvenDeterminer } from './odd-even-determiner';

export class Collatz {
    private collatzDataModel: CollatzDataModel;
    private oddEvenDeterminer: OddEvenDeterminer;

    constructor(startingNumber: number) {
        this.collatzDataModel = new CollatzDataModel(startingNumber, 0);
        this.oddEvenDeterminer = new OddEvenDeterminer();
    }

    getCollatzInformation(startingNumber: number): CollatzDataModel {
        this.collatzDataModel.setCollatzNumber(startingNumber);
        this.collatzDataModel.setCollatzIterations(this.getCollatzIterations(startingNumber));
        return CollatzDataModel.fromModel(this.collatzDataModel);
    }

    call(): CollatzDataModel {
        return this.getCollatzInformation(this.collatzDataModel.getCollatzNumber());
    }

    getCollatzIterations(startingNumber: number): number {
        let collatzIterations = 0;
        let collatzNumber = startingNumber;

        while (collatzNumber > 1) {
            if (this.oddEvenDeterminer.isOdd(collatzNumber)) {
                collatzIterations++;
                collatzNumber = this.collatzOddIteration(collatzNumber);
            } else {
                collatzIterations++;
                collatzNumber = this.collatzEvenIteration(collatzNumber);
            }
        }

        return collatzIterations;
    }

    collatzOddIteration(collatzNumber: number): number {
        return (collatzNumber * 3) + 1;
    }

    collatzEvenIteration(collatzNumber: number): number {
        return Math.floor(collatzNumber / 2);
    }
}
